src/btree/bt_curnext.c

```diff
@@ -104,27 +103,40 @@ __cursor_fix_next(WT_CURSOR_BTREE *cbt, bool newpage, bool restart)
     __cursor_set_recno(cbt, cbt->recno + 1);
 
 new_page:
+restart_read:
+    /* We only have one slot. */
+    cbt->slot = 0;
+
     /* Check any insert list for a matching record. */
     cbt->ins_head = WT_COL_UPDATE_SINGLE(page);
     cbt->ins = __col_insert_search(cbt->ins_head, cbt->ins_stack, cbt->next_stack, cbt->recno);
     if (cbt->ins != NULL && cbt->recno != WT_INSERT_RECNO(cbt->ins))
         cbt->ins = NULL;
-    /*
-     * FIXME-WT-6127: Now we only do transaction read if we have an update chain and it doesn't work
-     * in durable history. Review this when we have a plan for fixed-length column store.
-     */
     __wt_upd_value_clear(cbt->upd_value);
     if (cbt->ins != NULL)
-restart_read:
-        WT_RET(__wt_txn_read(session, cbt, NULL, cbt->recno, cbt->ins->upd));
+        /* Check the update list. */
+        WT_RET(__wt_txn_read_upd_list(session, cbt, cbt->ins->upd));
+    if (cbt->upd_value->type == WT_UPDATE_INVALID)
+        /* Nope. Read the on-disk value and/or history. */
+        WT_RET(__wt_txn_read(session, cbt, NULL, cbt->recno, cbt->ins ? cbt->ins->upd : NULL));
     if (cbt->upd_value->type == WT_UPDATE_INVALID) {
-        cbt->v = __bit_getv_recno(cbt->ref, cbt->recno, btree->bitcnt);
+        /*
+         * Getting here means that there was no update, the on-disk value isn't visible, and there
+         * isn't anything in history either. This means this chunk of the tree didn't exist yet for
+         * us (given our read timestamp), so we can either return NOTFOUND or produce a zero value
+         * depending on the desired end-of-tree semantics.
+         */
+        cbt->v = 0;
         cbt->iface.value.data = &cbt->v;
-    } else if (cbt->upd_value->type == WT_UPDATE_TOMBSTONE)
-        cbt->iface.value.data = cbt->upd_value->buf.data;
-    else
-        WT_RET(__wt_value_return(cbt, cbt->upd_value));
-    cbt->iface.value.size = 1;
+        cbt->iface.value.size = 1;
+    } else if (cbt->upd_value->type == WT_UPDATE_TOMBSTONE) {
```

Comment:
Keith also marked this; I wanted to keep them separate because one of the things coming soon is to see if we can make the end-of-tree location update transactionally instead of drifting around, and they're different paths then (the UPD_INVALID case becomes NOTFOUND) but I guess I should merge them for now. Fixed.